Add form for developers: "Request a new finder"

ChrisBAshton opened the pull request:
This adds a new form, aimed at developers (and protected behind `gds_admin` permission), for generating a schema for a new finder. It is accessible at `/admin/new`.

The form reuses the view partials, and the controller logic, from the existing 'edit' forms for metadata and facets, with minimal tweaks. The form can be submitted multiple times, and the POSTed values are retained, so the schema can be gradually built up over several submissions rather than requiring everything all in one go.

While we're here, we've also tidied up some calls to Design System margin classnames, and opened a can of worms RE the permissions / policy implementation.

Trello: https://trello.com/c/UpRa6hBs/3359-build-request-a-new-finder-form

## Screenshots

<table>
<tr><th>Landing page form</th><th>Generated schema page (auto-expanded)</th></tr>
<tr><td valign="top"><img src="https://github.com/user-attachments/assets/40f7fe4b-1bad-4e15-a99c-80c2dc097c25" alt=""></td><td valign="top"><img src="https://github.com/user-attachments/assets/3c6c1811-443b-49be-98e2-77736d96086f" alt=""></td></tr>
</table>

---

⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

Follow [these steps](https://guides.rubyonrails.org/upgrading_ruby_on_rails.html) if you are doing a Rails upgrade.

ChrisBAshton committed:
Add 'request new finder' route

ChrisBAshton committed:
Extract form fields into shared partials

In one of the next commits, we'll reuse these partials in the "Request
a new finder" form. We want just the raw inputs (and any
contextual guidance), without wrapping in a form or adding a
submit button (which will be handled on a per-form basis).

ChrisBAshton committed:
Move schema-overwriting parameter logic into shared methods

These methods are about to be reused by the `new` POST action.
This also makes the code simpler to read and follow.

ChrisBAshton committed:
Pull in styles for heading component

We'll use this on the new form

ChrisBAshton committed:
Standardise where margin-bottom/top is defined

We don't want to repeat this when reusing the partials on the
new form.

ChrisBAshton committed:
Add the form and POST route

ChrisBAshton committed:
Delete unused AttachmentPolicy

The "FIXME: this file is not currently used" comment has been in
place for [nine years](https://github.com/alphagov/specialist-publisher/commit/b6ecd6e46a2d95a2d26e3efc8e9e3e142012c47a#diff-40339c17ce537e6e8d70db8db23944a5f4a55d2a664146d6e4844de651940879R1).
It's not being fixed! Nor has it ever been known to cause a
problem.

In practice, the unused policy was configured to allow updates
from users who have gds_editor, departmental_editor or writer
permissions.

DocumentPolicy is being used instead of AttachmentPolicy, and that
policy already has the same permissions, except for the addition
of `document_type_editor`, which is a kind of cross-organisational
permission for folks to be able to edit documents for finders that
are not owned by the organisation they are a part of.
https://github.com/alphagov/specialist-publisher/blob/5c0cb882d5e27f0e8979d1433a861491e160e486/app/policies/application_policy.rb#L24-L26

TLDR, the DocumentPolicy permissions are fine for attachments too
and we don't need the complexity/overhead of breaking it down any
further than that.

ChrisBAshton committed:
Add TODO note about moving the remaining finder edit auth methods

These shouldn't really have ever been bundled into the
DocumentPolicy, since they refer to the policies around editing
the finder itself, not the documents associated with the finder.

Fixing these feels somewhat out of scope for this PR.

ryanb-gds commented on app/controllers/admin_controller.rb:
```diff
@@ -3,48 +3,37 @@ class AdminController < ApplicationController
+    @proposed_schema = FinderSchema.new
+    @proposed_schema.facets = []
+
+    if request.method == "POST"
```
Is there a reason we're not going for a standard `create` controller method?

ChrisBAshton replied:
Because I don't know Rails as well as I'd like 😁 I'll fix this up now.

ryanb-gds commented on app/policies/finder_administration_policy.rb:
```diff
@@ -0,0 +1,16 @@
+  # TODO: figure out why the exact same user object is passed in as
+  # two params from the controller. I must be doing something wrong,
+  # even if it's basically harmless!
+  def initialize(user, _user)
```
The Pundit gem's docs suggest you don't need to explicitly pass a user in when you call `authorize` in the controller, it passes the current_user in magically. That's why I'd guess you're getting double.

ChrisBAshton replied:
True... the [docs say](https://github.com/varvet/pundit#:~:text=The%20first%20argument,be%20anything%20really.):

> The first argument is a user. In your controller, Pundit will call the `current_user` method to retrieve what to send into this argument
> The second argument is some kind of model object, whose authorization you want to check. This does not need to be an ActiveRecord or even an ActiveModel object, it can be anything really.

I'm currently calling this in the controller:

```ruby
authorize current_user, :can_request_new_finder?, policy_class: FinderAdministrationPolicy
```

If I just drop `current_user` from that call, I get `undefined method new?' for an instance of FinderAdministrationPolicy` - so the `can_request_new_finder?` override seems to be ignored. If I go ahead and rename `def can_request_new_finder?` to `def new?`, it works, but then I'm needlessly passing a `:can_request_new_finder?` override (which is coming through as the second parameter - `_user`, in the code above). If I remove that override altogether (so now just `authorize policy_class: FinderAdministrationPolicy`) I get `ArgumentError in AdminController#new: wrong number of arguments (given 0, expected 1..2)`...

...it becomes a bit whackamole and I end up no further along - I still end up defining an unused parameter to the `initialize` method to get things working 🤔 I'll leave this for a future dev to sort out I think!

ryanb-gds replied:
Weird 🤷

ryanb-gds commented on app/views/admin/_facets_form_fields.html.erb:
```diff
@@ -0,0 +1,15 @@
+<p class="govuk-body">
+  Filters and its options are to help users find desired content on a specialist finder. These are selected by you for each specialist content item created in the Specialist Publisher. These are shown in the blue box on the documents and elsewhere.
```
The first sentence probably doesn't need "are to" in it?

ryanb-gds reviewed (approved):
This looks good, just a couple of observations

ChrisBAshton committed:
Swap POST route out for standard Rails 'create' method

This led to some duplication in the setup of the `proposed_schema`
variable but led to an opportunity to DRY up and simplify the
authorisation logic.
